packages/eslint-plugin/src/rules/restrict-template-expressions.ts

```diff
@@ -12,16 +12,35 @@ import {
   isTypeNeverType,
 } from '../util';
 
-type Options = [
-  {
-    allowAny?: boolean;
-    allowBoolean?: boolean;
-    allowNullish?: boolean;
-    allowNumber?: boolean;
-    allowRegExp?: boolean;
-    allowNever?: boolean;
+const optionsObj = {
+  Any: isTypeAnyType,
+  Array: (type, checker, rec): boolean => {
+    if (!checker.isArrayType(type)) {
+      return false;
+    }
+    const numberIndexType = type.getNumberIndexType();
+    return !numberIndexType || rec(numberIndexType);
   },
-];
+  Boolean: (type): boolean => isTypeFlagSet(type, ts.TypeFlags.BooleanLike),
+  Nullish: (type): boolean =>
+    isTypeFlagSet(type, ts.TypeFlags.Null | ts.TypeFlags.Undefined),
+  Number: (type): boolean =>
+    isTypeFlagSet(type, ts.TypeFlags.NumberLike | ts.TypeFlags.BigIntLike),
+  RegExp: (type, checker): boolean => getTypeName(checker, type) === 'RegExp',
+  Never: isTypeNeverType,
+} satisfies Record<
+  string,
+  (
+    type: ts.Type,
+    checker: ts.TypeChecker,
+    rec: (type: ts.Type) => boolean,
+  ) => boolean
+>;
+const optionsArr = Object.entries(optionsObj).map(([_type, fn]) => {
+  const type = _type as keyof typeof optionsObj;
+  return { type, option: `allow${type}` as const, fn };
+});
```

Comment:
[Refactor] This doesn't need both an `optionsObj` and an `optionsArr`. `optionsObj` is only ever used to create `optionsArr` - so we can inline them a bit.

```ts
const optionsTesters = [
  ["allowAny", isTypeAnyType],
  // ...
] as const satisfies [string, OptionTypeTester][];

type Options = [{ [Type in (typeof optionsTesters)[number][0]]?: boolean }];
```

...and similar changes lower down.